src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/PreserveApplications.cs

```diff
@@ -12,26 +12,64 @@
 
 namespace MonoDroid.Tuner {
 
-	public class PreserveApplications : BaseSubStep {
-
+	public class PreserveApplications
+#if NET5_LINKER
+	: IMarkHandler
+#else
+	: BaseSubStep
+#endif
+	{
+
+#if NET5_LINKER
+		LinkContext context;
+		AnnotationStore Annotations => context?.Annotations;
+
+		public void Initialize (LinkContext context, MarkContext markContext)
```

Comment:
Given all the new `Initialize()` additions, *and* that these are all very similar, and that all the *removals* are for `BaseStep`/`BaseSubStep`, might it be "cleaner" -- in terms of `diff` size -- to *re-introduce* these types?

```csharp
#if NET5_LINKER

public class BaseStep : IMarkHandler {
    LinkContext context;
    protected AnnotationStore Annotations => context?.Annotations;

    public virtual void Initialize (LinkContext context, MarkContext markContext)
    {
        this.context = context;
    }
}

public class BaseSubStep : BaseStep {
}

#endif  // !NET5_LINKER
```

This would allow us to have:

```csharp
public class PreserveApplications : BaseSubStep // no change!
{
    // .NET 5 specific; just override Initialize()
#if NET5_LINKER
    public override void Initialize (LinkContext context, MarkContext markContext)
    {
        base.Initialize (context, markContext);
        markContext.RegisterMarkAssemblyAction(…);
        markContext.RegisterMarkTypeAction(…);
    }
#endif  // !NET5_LINKER
}
```

Reply:
I addressed part of this by adding a `BaseMarkHandler` class with the shared members. Some of the .NET6 steps still need to be old-style `BaseSubStep`s though - for example `ApplyPreserveAttribute`. To merge the base classes, the new `BaseSubStep` would need to implement `ISubStep` and `IMarkHandler`This would let us get rid of this pattern in a lot of places:
```csharp
#if !NET5_LINKER
	override
#endif  // !NET5_LINKER
```

But there are also some downsides:
- it would duplicate https://github.com/mono/linker/blob/main/src/linker/Linker.Steps/BaseSubStep.cs
- it will be less obvious whether a given derived type of `BaseSubStep` is behaving like an `ISubStep` or an `IMarkHandler` (it depends how it is set up in the pipeline)

Let me know what you think!

Also, the old `BaseStep` is still used (for example `AddKeepAlivesStep`), and it can't implement both `IStep` and `IMarkHandler` otherwise the linker won't know which interface to use for the `--custom-step` argument (I think the failure mode is that it would treat all `BaseStep`s as `IMarkHandler`s). So I would leave `BaseStep` as-is, but it's worth considering for `BaseSubStep`.

Reply:
@sbomer: I like `BaseMarkHandler` as a compromise.

That said, I'm still not fully clear on everything.  Why does `ApplyPreserveAttribute` need to be an old-style `BaseSubStep`?  Presumably because of `PreserveSubStepDispatcher`'s `new ISubStep[]{new ApplyPreserveAttribute()}`, but…why is `PreserveSubStepDispatcher` required?  Why can't `ApplyPreserveAttribute` be an `IMarkHandler`?

Ditto `AddKeepAlivesStep`.

Related: what is the "long term evolution" of mono/linker?  Will `ISubStep` & related stick around, or are those going away?

Why can't both `IStep` and `IMarkHandler` be implemented?  You mention:

> the linker won't know which interface to use for the `--custom-step` argument (I think the failure mode is that it would treat all `BaseStep`s as `IMarkHandler`s).

(1) *Is* that a failure mode?  It sounds useful.  (2) How is that a problem?

Another problem for me is that I don't understand the difference between `IStep` and `IMarkHandler`; I *thought* `IMarkHandler` was the replacement for `IStep`/etc., so if `IMarkHandler` is invoked, and we can use `IMarkHandler` (…can we? See also above questions about `ApplyPreserveAttribute` & `AddKeepAlivesStep`), then I don't see an actual problem here.

---

I suspect that the "answer" to my questions lies in the original PR message:

> The exceptions are:
> 
> * ApplyPreserveAttribute: …

The problem is that I don't understand the underlying context.  (Docs? :smile:). mono/linker#1774 mentions "run logic on mark".  How many "steps" are there in the linker, what is their ordering, etc.? When is "mark" vs. "IStep", and how do they relate?  Why *can't* "everything" become IMarkHandler?  Why are there exceptions?

I don't understand. :frowning_face:

Reply:
> I don't understand the difference between IStep and IMarkHandler; I thought IMarkHandler was the replacement for IStep

The linker will call one `IStep` at a time, in the pipeline order, each operating on the set of assemblies loaded so far. `IMarkHandler` instead can register actions that get called, during MarkStep, whenever certain things are marked - it's designed for logic that should run over marked things instead of the entire set of loaded assemblies.

> Why does ApplyPreserveAttribute need to be an old-style BaseSubStep? [...] Why can't ApplyPreserveAttribute be an IMarkHandler?

That step needs to scan code for attributes, including _unmarked_ code. So as a compromise, the proposed behavior is to scan for attributes in all marked assemblies. To do this it could be an `IMarkHandler` but it would have to contain the logic to scan over all members in an assembly. That's where the `PreserveSubStepDispatcher` comes in. It's derived from [`MarkSubStepsDispatcher`](https://github.com/mono/linker/blob/8f1fc2560ce846fb4b8c85fc398d054ba58039ab/src/linker/Linker.Steps/MarkSubStepsDispatcher.cs#L13-L19) which handles the scanning part, and allows substeps including `ApplyPreserveAttribute` to only contain the logic to process members in the marked assembly.

> Ditto AddKeepAlivesStep.

This one isn't triggered by marking - instead it runs for everything that's left over after the Mark+Sweep, so it's just a plain old step:
https://github.com/xamarin/xamarin-android/blob/cf84e1b9a172a1e31d7eabf2f5eecd57ab57f301/src/Microsoft.Android.Sdk.ILLink/SetupStep.cs#L66

> what is the "long term evolution" of mono/linker? Will ISubStep & related stick around, or are those going away?

I don't see any reason for `ISubStep`/`IStep` to go away; they're still used in lots of places, here and in xamarin-macios. I might not be the best person to comment on the _long_ term evolution, but for .NET 6 our main goal has been to provide correctness guarantees with static analysis warnings. The new APIs make it possible to correctly handle assemblies that are dynamically loaded (discovered by the linker only during MarkStep), and also allow custom steps to avoid scanning the world for operations that only need to touch marked things. /cc @marek-safar

> Why can't both IStep and IMarkHandler be implemented?

>> I think the failure mode is that it would treat all BaseSteps as IMarkHandlers

> (1) Is that a failure mode? It sounds useful.

The old-style BaseSteps weren't designed to be triggered by mark, and probably wouldn't do the right thing - `IStep` is a different beast than `IMarkHandler`. If there are any specific old-style BaseSteps that could be changed to run for marked members, we could make them `IMarkHandlers`, but any that need to go after CleanStep would be unchanged.

> How many "steps" are there in the linker, what is their ordering

This in particular is something I've been trying to untangle - the ordering is important in a lot of cases but it's not obvious when. This comment should give the current ordering for built-in steps: https://github.com/mono/linker/blob/main/src/linker/Linker/Driver.cs#L714.

I hope that helps a little bit - I'll update the linker docs to make some of this easier to discover in the future.